File: algo_2021/w5.-traverse-divide-and-conquer/problems/480.Binary_Tree_Paths/480_binary_tree_paths__div_conq.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: the root of the binary tree
    @return: all root-to-leaf paths
    """

    # divide conquer solution 
    def binaryTreePaths(self, root):
        # write your code here
        res = []
        if not root:
            return res
        if not root.left and not root.right:
            return [str(root.val)]
        for path in self.binaryTreePaths(root.left):
            res.append(str(root.val) + '->' + path)
        for path in self.binaryTreePaths(root.right):
            res.append(str(root.val) + '->' + path)
        return res

File: lintcode_algo_2021/s5.-traverse-d-and-c/480.Binary_Tree_Paths/480_binary_tree_paths__div_conq.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */

public class Solution {
    /**
     * @param root: the root of the binary tree
     * @return: all root-to-leaf paths
     */
    public List<String> binaryTreePaths(TreeNode root) {
        // write your code here
        List<String> res = new ArrayList<>();
        if (root == null) {
            return res;
        }
        if (root.left == null && root.right == null) {
            res.add("" + root.val);
            return res;
        }

        for (String leftPath : binaryTreePaths(root.left)) {
            res.add(root.val + "->" + leftPath);
        }
        for (String rightPath : binaryTreePaths(root.right)) {
            res.add(root.val + "->" + rightPath);
        }
        
        return res;
    }
}